### PR TITLE
itest: add new testCustomChannelsHtlcSuccess itest

### PR DESCRIPTION
**Pull request by Roasbeef**

### CODE REVIEW

**Comment by guggero** [comment]
Pushed up a rebase.

**Comment by Roasbeef** [comment]
itests are racy right now due to sweeper/block timing, experimenting with either using `BumpFee` to trigger an immediate sweep or, be more lax w.r.t asserting _which_ block mines the target txn.

**Comment by guggero** [comment]
Pushed up a rebase to reference the latest version of https://github.com/lightninglabs/taproot-assets/pull/1154.

**Comment by Roasbeef** [comment]
Updated to ref the latest `tapd`+`lnd`.

File: itest/assets_test.go [hunk 1]
```diff
@@ -1576,6 +1578,70 @@ func assertAssetBalance(t *testing.T, client *tapClient, assetID []byte,
...
+
+		var assetSum uint64
+		for _, asset := range relevantAssets {
+			if asset.ScriptKeyIsLocal {
+				assetSum += asset.Amount
+			}
+		}
+
+		if assetSum != expectedBalance {
+			return fmt.Errorf("expected balance %d, got %d", expectedBalance, assetSum)
```
Comment (guggero) [inline itest/assets_test.go:1700]: nit: lots of "line too long" in these commits... Did you change editor by any chance?

File: itest/assets_test.go [hunk 2]
```diff
@@ -786,7 +786,8 @@ func createAndPayNormalInvoice(t *testing.T, src, rfqPeer, dst *HarnessNode,
 	require.NoError(t, err)
 
 	numUnits, _ := payInvoiceWithAssets(
-		t, src, rfqPeer, invoiceResp, assetID, smallShards,
+		t, src, rfqPeer, invoiceResp.PaymentRequest, assetID, smallShards,
+		fn.None[lnrpc.Payment_PaymentStatus](),
```
Comment (guggero) [inline itest/assets_test.go:790]: nit: could add a `defaultPaymentStatus = fn.None[lnrpc.Payment_PaymentStatus]()` at the top of the file to shorten this across this commit.

File: itest/litd_custom_channels_test.go [hunk 3]
```diff
@@ -2766,3 +2771,457 @@ func testCustomChannelsFee(_ context.Context,
 		"min_relay_fee: ", tooLowFeeRateAmount.FeePerKWeight())
 	require.ErrorContains(t.t, err, errFeeRateTooLow)
 }
+
+func assertNumHtlcs(t *testing.T, node *HarnessNode, expected int) {
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:2776]: nit: could move these helpers to the `assets_test.go` file to de-clutter the main itest file a bit.

File: itest/litd_custom_channels_test.go [hunk 4]
```diff
@@ -2766,3 +2771,457 @@ func testCustomChannelsFee(_ context.Context,
...
+		t.t, "Alice", lndArgs, false, true, litdArgs...,
+	)
+	require.NoError(t.t, err)
+	bob, err := net.NewNode(
+		t.t, "Bob", lndArgs, false, true, litdArgs...,
+	)
+	require.NoError(t.t, err)
+
+	//fmt.Println("created alice and bob")
+	//time.Sleep(time.Second * 10)
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:2893]: nit: can remove?

File: itest/litd_custom_channels_test.go [hunk 5]
```diff
@@ -2766,3 +2771,457 @@ func testCustomChannelsFee(_ context.Context,
...
+	mineBlocks(t, net, 1, 1)
+
+	// TODO(roasbeef): assert transfers
+
+	t.Logf("Confirming Bob's remote HTLC success sweep")
+
+	// Bob's balance should now reflect that he's gained the value of the HTLC,
+	// in addition to his settled balance.
+	//
+	// TODO(roasbeef): -1 as no decimal display on asset?
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:3090]: Yeah, that seems to be the likely cause.

File: itest/litd_custom_channels_test.go [hunk 6]
```diff
@@ -2766,3 +2771,457 @@ func testCustomChannelsFee(_ context.Context,
...
+	// Bob's incoming HTLC sweep directly off the commitment transaction.
+	_, err = waitForNTxsInMempool(
+		net.Miner.Client, 1, time.Second*5,
+	)
+
+	// We'll now mine the next block, which should confirm Bob's HTLC sweep
+	// transaction.
+	mineBlocks(t, net, 1, 1)
+
+	// TODO(roasbeef): assert transfers
```
Comment (guggero) [inline itest/litd_custom_channels_test.go:3083]: Asserting that the correct proofs can be found in the universe would also be a great check.

File: itest/litd_test_list_on_test.go [hunk 7]
```diff
@@ -48,6 +48,10 @@ var allTestCases = []*testCase{
 		name: "test custom channels liquidity",
 		test: testCustomChannelsLiquidityEdgeCases,
 	},
+	{
+		name: "test custom channels htlc success",
```
Comment (guggero) [inline itest/litd_test_list_on_test.go:52]: I guess we'll also want to add a test case for timing out the HTLCs, instead of resolving them?

**Review by guggero — COMMENTED** [review]
Nice! Wasn't able to get the test to run on my machine even after many tries. So perhaps this isn't the latest version yet?

File: itest/litd_test_list_on_test.go (hunk 7 above)
Comment (Roasbeef) [reply itest/litd_test_list_on_test.go:52]: Bad test name here. We timeout out HTLCs at the very end of the itest when we mine enough blocks to timeout the HTLCs.

File: itest/litd_custom_channels_test.go (hunk 4 above)
Comment (Roasbeef) [reply itest/litd_custom_channels_test.go:2893]: Yep, stray left over to find the PID to attach the debugger to. Later on I added the change to print the PID.

File: itest/litd_test_list_on_test.go (hunk 7 above)
Comment (Roasbeef) [reply itest/litd_test_list_on_test.go:52]: So this tests all 4 scenarios we about ( (local, remote) x (success, timeout) ), it does it all in one sweep with Alice's commitment that she force closed. I'll change the name of the test to better reflect what's covered.

**Comment by Roasbeef** [comment]
Pushed up a new version!

Ran this for 30 minutes on my laptop and didn't get a flake. If it flakes on CI, I should be able to address it easily, given all the quality time spent with this test. 

I also added a final assertion to make sure both Alice and Bob can send all the swept funds to a 3rd party.

**Comment by guggero** [comment]
Pushed up a rebase.

File: itest/litd_custom_channels_test.go [hunk 8]
```diff
@@ -2768,3 +2769,432 @@ func testCustomChannelsFee(_ context.Context,
...
+		t.t, aliceTap, assetID, aliceExpectedBalance,
+	)
+
+	t.Logf("Sending all settled funds to Zane")
+
+	// As a final sanity check, both Alice and Bob should be able to send their
+	// entire balances to Zane, our 3rd party.
+	//
+	// We'll make two addrs for Zane, one for Alice, and one for bob.
+	zaneTap := newTapClient(t.t, zane)
```
Comment (GeorgeTsagk) [inline itest/litd_custom_channels_test.go:3175]: if we keep this we can remove the commented out zane tap init on top of the test

File: itest/litd_custom_channels_test.go [hunk 9]
```diff
@@ -2768,3 +2769,432 @@ func testCustomChannelsFee(_ context.Context,
...
+		t.t, "Zane", lndArgs, false, true, litdArgs...,
+	)
+	require.NoError(t.t, err)
+
+	litdArgs = append(litdArgs, fmt.Sprintf(
+		"--taproot-assets.proofcourieraddr=%s://%s",
+		proof.UniverseRpcCourierType, zane.Cfg.LitAddr(),
+	))
+
+	// Next, we'll make Alice and Bob, who will be the main nodes under test.
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply itest/litd_custom_channels_test.go]: Old diff? Looks fine to me

File: itest/litd_custom_channels_test.go [hunk 10]
```diff
@@ -2768,3 +2769,432 @@ func testCustomChannelsFee(_ context.Context,
...
+		t.t, "Bob", lndArgs, false, true, litdArgs...,
+	)
+	require.NoError(t.t, err)
+
+	// Now we'll connect all nodes, and also fund them with some coins.
+	nodes := []*HarnessNode{alice, bob}
+	connectAllNodes(t.t, net, nodes)
+	fundAllNodes(t.t, net, nodes)
+
+	//universeTap := newTapClient(t.t, zane)
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply itest/litd_custom_channels_test.go:2809]: Removed for now. It's all HTLC sweeps so nothing that a party other than the sweeper need to be aware of. We have existing coverage for the force close and funding txns in the other test.

File: itest/litd_custom_channels_test.go (hunk 8 above)
Comment (Roasbeef) [reply itest/litd_custom_channels_test.go:3175]: Not sure what you mean. Zane is the universe server that information is posted to, then we use it at the very end to ensure that we can send the funds we've swept.

**Review by guggero — APPROVED** [review]
Gonna merge then absorb the commits into the `update-to-lnd-18-4` branch, fixing any nits along the way.

File: itest/litd_custom_channels_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go]: maybe, I think there was a fix-up between comment & clicking the "finish review" button

**Review by GeorgeTsagk — COMMENTED** [review]
post-merge ack

**Pull request merged**